### DISCUSSION TITLE
Assertion on 'speed' failed when building emission model

### DISCUSSION THREAD

**Discussion by nadiaqamilla** [ipeaGIT/gtfs2emis#102] [Q&A] [answered]
Hi, i tried to use the package to build emission model using Transjakarta GTFS (BRT in Indonesia). 
After building the transport model using code below:
`  tp_model <- gtfs2emis::transport_model(gtfs_data = gtfs,
                              spatial_resolution = 100,
                              parallel = TRUE) `
It shoes warnings below in the console:
**30450 'speed' values are Inf for shapeid 'xwq4'.                                                                                                   
There were 15 warnings (use warnings() to see them)**

Afterwards i generated emission model using code below:
`emi_list <- emission_model(tp_model = tp_model,
                           ef_model = "ef_europe_emep",
                           fleet_data = fleet_df,
                           pollutant = c("CO2","PM10"),
                           reference_year = 2020
)`

The code resulted in error below:
**Error in ef_europe_emep(pollutant = pollutant, speed = temp_shape$speed,  : 
  Assertion on 'speed' failed: Contains missing values (element 1576024).**

Any help would be really appreciated. Thank you

**Comment by Joaobazzo** [comment, marked as answer]
Hi @nadiaqamilla ,
Thanks for bringing this up. I don't have the Transjakarta GTFS to reproduce the results, but it seems a familiar issue. This probably happens because there is an invalid speed being input into the emissions model (which could be NA, Inf, or NaN). One possible workaround is fixing the speed of the shape_id `xwq4`. In the example below, I changed the speed values to 36 km/h.  Let me know if this works for you.

```R
 library(gtfstools)
library(data.table)
library(gtfs2gps)
library(magrittr)
library(gtfs2emis)

# run transport model (for only shape_id "xwq4")
tp_model_fix_speed <- gtfs_data %>% 
  gtfstools::filter_by_shape_id("xwq4") %>% 
  gtfs2gps::gtfs2gps() %>% 
  gtfs2gps::adjust_speed(new_speed = units::set_units(36,"km/h")) %>% 
  gtfs2gps::gps_as_sflinestring()

# emissions from fix shape_id
emi_list_fix <- emission_model(tp_model = tp_model_fix_speed,
                               ef_model = "ef_europe_emep"
                           , fleet_data = fleet_df,
                           pollutant = c("CO2","PM10")
                           , reference_year = 2020 )

# run transport model (for all shape_id`s except "xwq4")

all_shapeids <- setdiff(unique(gtfs_small$shapes$shape_id),"xwq4")
new_gtfs <- gtfstools::filter_by_shape_id(gtfs_small,all_shapeids)
new_tp_model <- gtfs2emis::transport_model(gtfs_data = new_gtfs
                                       , spatial_resolution = 100,
                                       parallel = TRUE)

# emissions model (for all shape_id's except "xwq4")

new_emi_list <- emission_model(tp_model = new_tp_model
                           , ef_model = "ef_europe_emep"
                           , fleet_data = fleet_df
                           , pollutant = c("CO2","PM10")
                           , reference_year = 2020 )


```
